src/iterative_ensemble_smoother/sies.py

```diff
@@ -0,0 +1,416 @@
+        H = S @ self.W + self.D - g_X
+
+        # Line 8
+        # Some asserts to remind us what the shapes of the matrices are
+        assert self.W.shape == (N, N)
+        assert S.shape == (m, N)
+        assert self.C_dd.shape in [(m, m), (m,)]
+        assert H.shape == (m, N)
+
+        proposed_W: npt.NDArray[np.double] = self.inversion(
+            W=self.W,
+            step_length=step_length,
+            S=S,
+            C_dd=self.C_dd,
+            H=H,
+            C_dd_cholesky=self.C_dd_cholesky,
+            truncation=self.truncation,
+        )
+        return proposed_W
+
+    def propose_W_masked(
+        self,
+        responses: npt.NDArray[np.double],
+        ensemble_mask: npt.NDArray[np.bool_],
+        step_length: float = 0.5,
+    ) -> npt.NDArray[np.double]:
+        """Returns a proposal for W_i, without updating the internal W.
+
+        This is an implementation of lines 4-8 in Algorithm 1.
+
+        Parameters
+        ----------
+        responses : npt.NDArray[np.double]
+            The model evaluated at X_i. In other words, responses = g(X_i).
+            This is Y in the paper.
+        step_length : float, optional
+            Step length for Gauss-Newton. The default is 0.5.
+
+        Returns
+        -------
+        W_i : npt.NDArray[np.double]
+            A proposal for a new W in the algorithm.
+
+        """
+        Y = responses
+        assert Y.ndim == 2
+        assert Y.shape[0] == self.C_dd.shape[0]
+        g_X = Y.copy()
+        if ensemble_mask is not None:
+            assert Y.shape[1] == ensemble_mask.sum()
+
+        # Get shapes. Same notation as used in the paper.
+        N = self.X.shape[1]  # Ensemble members in prior
+        n = self.X.shape[0]  # Parameters (inputs)
+        m = self.C_dd.shape[0]  # Responses (outputs)
+        k = Y.shape[1]  # Active ensemble members
+
+        # Line 4 in Algorithm 1.
+        Y = (g_X - g_X.mean(axis=1, keepdims=True)) / np.sqrt(N - 1)
+
+        # Line 5
+        Omega = self.W.copy()
+        Omega -= Omega.mean(axis=1, keepdims=True)
+        Omega /= np.sqrt(N - 1)
+        Omega.flat[:: Omega.shape[0] + 1] += 1  # Add identity in place
+        Omega = Omega[:, ensemble_mask]
+
+        # Line 6
+        if n < k - 1:  # Here we use k instead of N
+            # There are fewer parameters than realizations. This means that the
+            # system of equations is overdetermined, and we must solve a least
+            # squares problem.
+
+            # An alternative approach to producing A_i would be keeping the
+            # returned value from the previous Newton iteration (call it X_i),
+            # then computing:
+            # A_i = (self.X_i - self.X_i.mean(axis=1, keepdims=True)) / np.sqrt(N - 1)
+            A_i = self.A @ Omega
+            # S = sp.linalg.solve(
+            #     Omega.T, np.linalg.multi_dot([Y, sp.linalg.pinv(A_i), A_i]).T
+            # ).T
+            ST, *_ = sp.linalg.lstsq(
```

Comment:
Why not use `sp.linalg.solve` here?

Reply:
i think there might have been a bug earlier where we said that `W` had to be `W[mask, mask]`. but if we start out with e.g. 10, and we end up with 5, that does not mean that the remaining 5 have to be linear combinations of the ORIGINAL 5. I think `W[:, mask]` is correct.

Reply:
i think there might have been a bug earlier where we said that `W` had to be `W[mask, mask]`. but if we start out with e.g. 10, and we end up with 5, that does not mean that the remaining 5 have to be linear combinations of the ORIGINAL 5. I think `W[:, mask]` is correct.